gumby/scenario.py

```diff
@@ -87,59 +85,75 @@ class ScenarioParser():
                time stamp, they will be executed in order.
     """
     _re_substitution = re_compile("(\$\w+)")
+    _re_preprocessor_dir = re_compile("^#(\w+)\s+")
+    _re_named_arg = re_compile("^\s*(\w+)\s*=\s*(.*)$")
 
     def __init__(self):
+        super(ScenarioParser, self).__init__()
         self._logger = logging.getLogger(self.__class__.__name__)
         self.file_lock = RLock()
-        self.file_buffer = None
+        self.line_buffer = []
+        self.preprocessor_callbacks = {
+            "include": self._preproc_include_file
+        }
 
-    def _read_scenario(self, filename):
+    def add_scenario(self, filename):
         """
-        Read the scenario into memory return a list containing the lines.
+        Read the scenario into this scenario parser
         """
         with self.file_lock:
-            if not self.file_buffer or self.file_buffer[0] != filename:
-                f = open(filename, "r")
-                lines = f.readlines()
-                f.close()
-
-                line_buffer = []
-
-                linenr = 1
-                for line in lines:
-                    if not line.startswith('#'):
-                        line = line.strip()
-                        line_buffer.append((linenr, line))
-                    linenr += 1
-
-                self.file_buffer = (filename, line_buffer)
-        return self.file_buffer[1]
-
-    def _parse_scenario(self, filename):
+            f = open(filename, "r")
+            lines = f.readlines()
+            f.close()
+
+            line_number = 1
+            for line in lines:
+                if not line.startswith('#'):
+                    line = line.strip()
+                    self.line_buffer.append((filename, line_number, line))
```

Comment:
I wanted to keep the filename with the line number. That way errors ("unregistered callback used", "module import error" or "include file doesn't exist") can tell you exactly where the problem is.